pkg/block/forest.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package block
+
+import (
+	"github.com/oklog/ulid"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+// Vertex type implements a vertex of a Forest.
+type Vertex struct {
+	ID       ulid.ULID
+	Value    *metadata.Meta
+	Children []*Vertex
+}
+
+// NewVertex creates a new vertex.
+func NewVertex(id ulid.ULID, value *metadata.Meta) *Vertex {
+	v := &Vertex{
+		ID:       id,
+		Children: []*Vertex{},
+		Value:    value,
+	}
+
+	return v
+}
+
+// AddMeta Inserts meta into forest such that the Sources of parent vertex is superset of all child vertices.
+func (f *Vertex) AddMeta(id ulid.ULID, metas map[ulid.ULID]*metadata.Meta) bool {
+	return f.add(id, f, metas)
+}
+
+// GetNonRootIDs returns list of ids which are not on root level of forest.
+func (f *Vertex) GetNonRootIDs() []ulid.ULID {
+	var ulids []ulid.ULID
+	for _, vertex := range f.Children {
+		ulids = append(ulids, f.iterate(vertex)...)
+		ulids = remove(ulids, vertex.ID)
+	}
+	return ulids
+}
+
+func (f *Vertex) iterate(vertex *Vertex) []ulid.ULID {
+	var ulids []ulid.ULID
+	ulids = append(ulids, vertex.ID)
+	for _, childVertex := range vertex.Children {
+		ulids = append(ulids, f.iterate(childVertex)...)
+	}
+	return ulids
+}
+
+func (f *Vertex) add(id ulid.ULID, startVertex *Vertex, metas map[ulid.ULID]*metadata.Meta) bool {
+	var rootVertex *Vertex
+	newVertex := NewVertex(id, metas[id])
+	for _, vertex := range startVertex.Children {
+		rootSources := metas[vertex.ID].Compaction.Sources
+		sources := metas[id].Compaction.Sources
+
+		if f.containsSources(rootSources, sources) && f.containsSources(sources, rootSources) { // a block exists with same sources.
+			vertex.Children = append(vertex.Children, newVertex) // add as child so it is deleted.
+			return true
+		}
+
+		if f.containsSources(rootSources, sources) {
+			rootVertex = vertex
+			break
+		}
+	}
+
+	if rootVertex == nil {
+		startVertex.Children = append(startVertex.Children, newVertex)
+		return true
+	}
+
+	return f.add(id, rootVertex, metas)
+}
+
+func (f *Vertex) containsSources(rootSources, sources []ulid.ULID) bool {
```

Comment:
Does it have to be method? Also, I don't get why we need this at all this method does nothing ): It might too shallow